test/mpi/comm/cmsplit_type.c

```diff
@@ -55,6 +55,63 @@ int main(int argc, char *argv[])
         MPI_Comm_free(&comm);
     }
 
+    /* Check to see if MPI_COMM_TYPE_HW_GUIDED works correctly */
+    for (i = 0; split_topo[i]; i++) {
+        MPI_Info_create(&info);
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[i]);
+        int ret;
+        ret = MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+        /* result will depend on platform and process bindings, just check returns */
+        if (ret != MPI_SUCCESS) {
+            printf("MPI_COMM_TYPE_HW_GUIDED (%s) failed\n", split_topo[i]);
+            errs++;
+        } else if (comm != MPI_COMM_NULL) {
+            MPI_Comm_free(&comm);
+        }
+        MPI_Info_free(&info);
+    }
+
+    /* Test MPI_COMM_TYPE_HW_GUIDED: pass MPI_INFO_NULL, it must return MPI_COMM_NULL. */
+    info = MPI_INFO_NULL;
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+    if (comm != MPI_COMM_NULL) {
+        printf("MPI_COMM_TYPE_HW_GUIDED with MPI_INFO_NULL didn't return MPI_COMM_NULL\n");
+        errs++;
+        MPI_Comm_free(&comm);
+    }
+
+    /* Test MPI_COMM_TYPE_HW_GUIDED: info without correct key, it must return MPI_COMM_NULL. */
+    MPI_Info_create(&info);
+    /* note: shmem_topo is the wrong key for MPI_COMM_TYPE_HW_GUIDED */
+    MPI_Info_set(info, "shmem_topo", split_topo[0]);
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+    if (comm != MPI_COMM_NULL) {
+        printf("MPI_COMM_TYPE_HW_GUIDED without correct key didn't return MPI_COMM_NULL\n");
+        errs++;
+        MPI_Comm_free(&comm);
+    }
+    MPI_Info_free(&info);
+
+    /* Check to see if MPI_COMM_TYPE_HW_UNGUIDED works correctly */
+    MPI_Info_create(&info);
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_UNGUIDED, 0, info, &comm);
+    if (comm != MPI_COMM_NULL) {
+        int newsize;
+        MPI_Comm_size(comm, &newsize);
+        if (!(newsize < world_size)) {
+            printf("MPI_COMM_TYPE_HW_UNGUIDED: Expected comm to be a proper sub communicator\n");
```

Comment:
Can you please explain why `new_size == world_size` is wrong? And maybe it is better to use `newsize >= world_size` rather than `!(<)`.

Reply:
`MPI_COMM_TYPE_HW_UNGUIDED` needs to return *strict* subset of original comm per specification, so it requires the new comm size strictly less than the original size. `newsize >= world_size` just read silly since how can the comm split resulting in a bigger comm than original?

Reply:
Do you want to keep the "!(newsize < world_size)" code?

Reply:
I am on the other side. Anyway I don't think this line is worth for us to spend more time on it. I am fine with it